### PR TITLE
Release 2.0.0

### PR DESCRIPTION
**Pull request by bytemain**
Breaking Changes:
- We remove the direct import of reflect-metadata
- Remove `__id` and `__injectorId` from the created instance

Feat:
- Support hook priority option
- Ship with both esm and commonjs dist
- Support factory/multiple instance check

fix: #111 
fix: #102 

---

Behind scene(May the users who do not care):

- The InstanceCreator method sign is changed:
```diff
interface InstanceCreator {
- instance: any;
+ instances: Set<any>;
}
```

### CODE REVIEW

File: src/injector.ts [hunk 1]
```diff
@@ -433,37 +447,29 @@ export class Injector {
...
 
-    if (Helper.isFactoryCreator(creator)) {
-      return applyHooks(creator.useFactory(this), token, this.hookStore);
+    if (isFactoryCreator(creator)) {
+      return this.createInstanceFromFactory(ctx as Context<FactoryCreator>);
     }
 
     // must be ValueCreator, no need to hook.
-    return creator.instance;
+    return creator.instances!.values().next().value;
```
Comment (erha19) [inline src/injector.ts:472]: 这里类型太严格，感觉可能会报错

File: test/decorators/hooks.test.ts [hunk 2]
```diff
@@ -394,6 +394,102 @@ describe('hook', () => {
     await Promise.all(pendings);
   });
 
+  describe('hook priority', () => {
+    it("hook's priority should work case1", async () => {
```
Comment (erha19) [inline test/decorators/hooks.test.ts:398]: 应该用单引号统一格式

**Review by erha19 — CHANGES_REQUESTED** [review]

File: test/decorators/hooks.test.ts [hunk 3]
```diff
@@ -394,6 +394,102 @@ describe('hook', () => {
...
+        },
+      ]);
+
+      const testClass = injector.get(TestClass);
+      expect(await testClass.add(1, 2)).toBe(10);
+      expect(resultArray).toEqual([0, 1, 2, 3]);
+    });
+  });
+
   it('子injector应该正确拦截', () => {
```
Comment (erha19) [inline test/decorators/hooks.test.ts:493]: 2.0 可以顺带优化一下这些 test case 命名

File: test/decorators/hooks.test.ts (hunk 2 above)
Comment (bytemain) [reply test/decorators/hooks.test.ts:398]: 这个字符串内有个单引号，所以 format 工具自动变成双引号了

**Comment by bytemain** [comment]
待补充文档：集成方需要自行引用 reflect metadata

**Review by erha19 — APPROVED** [review]
LGTM

**Pull request merged**